source/Timer.cpp

```diff
@@ -0,0 +1,253 @@
+				{Timer::ResetCondition::NONE, "none"},
+				{Timer::ResetCondition::LEAVE_SYSTEM, "leave system"},
+				{Timer::ResetCondition::LEAVE_ZONE, "leave zone"}
+			};
+			auto it = resets.find(resetCondition);
+			if(it != resets.end())
+				out.Write("reset", it->second);
+		}
+		if(proximity > 0.)
+		{
+			string distance = "close";
+			if(!closeTo)
+				distance = "far";
+			out.Write("proximity", proximity, distance);
+			if(proximityCenter != nullptr)
+			{
+				out.BeginChild();
+				{
+					out.Write(proximityCenter->Name());
+				}
+				out.EndChild();
+			}
+		}
+
+		action.Save(out);
+	}
+	out.EndChild();
+}
+
+// Calculate the total time to wait, including any random value,
+// and instantiate the triggered action
+Timer Timer::Instantiate(map<string, string> &subs,
+						const System *origin, int jumps, int64_t payload) const
+{
+	Timer result;
+	result.name = name;
+	result.base = base;
+	result.rand = rand;
+	result.requireIdle = requireIdle;
+	result.requireUncloaked = requireUncloaked;
+	result.system = system;
+	result.systems = systems;
+	result.proximity = proximity;
+	result.proximityCenter = proximityCenter;
+	result.closeTo = closeTo;
+	result.resetCondition = resetCondition;
+
+	result.timeToWait = timeToWait;
+	result.timeElapsed = timeElapsed;
+	result.isComplete = isComplete;
+	result.isActive = isActive;
+
+	result.action = action.Instantiate(subs, origin, jumps, payload);
+
+	return result;
+}
+
+bool Timer::IsComplete() const
+{
+	return isComplete;
+}
+
+void Timer::ResetOn(ResetCondition cond)
+{
+	bool reset = cond == resetCondition;
+	reset |= (cond == Timer::ResetCondition::LEAVE_ZONE && resetCondition == Timer::ResetCondition::PAUSE);
+	reset |= (cond == Timer::ResetCondition::LEAVE_SYSTEM && (resetCondition == Timer::ResetCondition::PAUSE
+				|| resetCondition == Timer::ResetCondition::LEAVE_ZONE));
+	if(isActive && reset)
+	{
+		timeElapsed = 0;
+		timeToWait = base + Random::Int(rand);
+	}
+}
+
+void Timer::Step(PlayerInfo &player, UI *ui)
+{
+	if(isComplete)
+		return;
+	if((system && player.Flagship()->GetSystem() != system) ||
+		(!systems.IsEmpty() && !systems.Matches(player.Flagship()->GetSystem())))
+	{
+		ResetOn(Timer::ResetCondition::LEAVE_SYSTEM);
+		isActive = false;
+		return;
+	}
+	if(requireIdle)
+	{
+		bool shipIdle = (!player.Flagship()->IsThrusting() && !player.Flagship()->IsSteering()
+						&& !player.Flagship()->IsReversing());
+		for(const Hardpoint &weapon : player.Flagship()->Weapons())
+			shipIdle &= !weapon.WasFiring();
+		if(!shipIdle)
+		{
+			ResetOn(Timer::ResetCondition::PAUSE);
+			isActive = false;
+			return;
+		}
+	}
+	if(requireUncloaked)
+	{
+		double cloak = player.Flagship()->Cloaking();
+		if(cloak != 0.)
+		{
+			ResetOn(Timer::ResetCondition::PAUSE);
+			isActive = false;
+			return;
+		}
+	}
+	if(proximity > 0.)
+	{
+		Point center;
+		if(proximityCenter)
+		{
+			const StellarObject *proximityObject = system->FindStellar(proximityCenter);
```

Comment:
A system might have multiple instances of the given planet, with the classic example being ringworlds. Presumably, it should apply to every single stellar object with the given planet, not just one.

Reply:
Ooh, I hadn't realized those were all instances of the same planet, though it makes sense. 

...So how does one go about getting _all_ the instances of a planet within a system? I'm not seeing a method on either System or Planet to do that. Is just iterating through `System::Objects()` the best way?

Reply:
> Is just iterating through System::Objects() the best way?

yup

Reply:
Got this implemented, and tested it at Alta Hai—all looked good.